### DISCUSSION TITLE
Mismatch between DH matrix computation and visualization of robot on webots

### DISCUSSION THREAD

**Discussion by alecive** [cyberbotics/webots#3695] [Q&A]
Hello, I'm a relatively new user of Webots and I trying to create a simple exercise in which students can understand Denavit Hartenberg. I'm using UR5e, and a modified version of the UR demo in ENU coordinates (I used the scripts in [here ](https://github.com/cyberbotics/webots/tree/master/scripts/converter) to modify the default world file to use ENU).

What I am doing is the following: 

1. I imported DH parameters of the UR5e from [here](https://www.universal-robots.com/articles/ur/application-installation/dh-parameters-for-calculations-of-kinematics-and-dynamics/)
2. I created my DH chain, computed the forward kinematics of the robot and checked that it matches with the excel file available from the link above and also [here](https://s3-eu-west-1.amazonaws.com/ur-support-site/45257/DH-Transformation.xlsx)
3. I computed the end-effector pose by pre-multiplying the forward kinematics above with the base of the robot
4. I visualized the pose with a Transform object

![image](https://user-images.githubusercontent.com/4378663/133687674-49e04586-9cdb-4bb9-8172-f9c95e565939.png)

Unfortunately, as you can see here the computed pose does not match with the end-effector pose. It seems to be an issue with rotations, however I do not understand how to debug this (primarily because I'm a new user):

- How can I get the position of the end-effector from within webots?
- Where can I see how webots rotates bodies and reference frames?
- Do you think there may be a hidden bug (for example, emerging from this PR https://github.com/cyberbotics/webots/pull/1643)?
- Is there a way to create a DH matrix from the proto file? I see there is a URDF-to-PROTO converter, but not the other way around.

Also, I'm trying not to share my code given that this is an assignment (but I can share bits of it if needed). Here is my world file: [ure.zip](https://github.com/cyberbotics/webots/files/7181105/ure.zip)

Thanks for the help!

**Comment by lukicdarkoo** [comment]
> How can I get the position of the end-effector from within webots?

Select the gripper and click on the `Position` tab in the scene tree:

![Screenshot-20210917092307-349x530](https://user-images.githubusercontent.com/2135826/133741766-f3bb9f1f-2c93-4c12-b612-113b90a2afa3.png)

> Where can I see how webots rotates bodies and reference frames?

You can right-click on the robot and `Convert to Base Node(s)` to display the full node tree in the scene tree. Then, in the scene tree, you can see a pose (relative to any other node), velocity, joint properties...

> Do you think there may be a hidden bug (for example, emerging from this PR Allow Z vertical axis #1643)?

Probably not as we didn't touch anything UR specific.

> Is there a way to create a DH matrix from the proto file? I see there is a URDF-to-PROTO converter, but not the other way around.

You can save it as a URDF. Right-click on the robot, `Export`, and save it as `<something>.urdf`.

**Comment by alecive** [comment]
Thanks for the feedback @lukicdarkoo . Here are my replies:

> Select the gripper and click on the Position tab in the scene tree:

I am aware of that. But how can I do that programmatically? I tried to `getField("Position").getSFVec3f()` but it doesn't work. Also, I am not understanding how to get the list of attributes for a node/object.

> You can right-click on the robot and Convert to Base Node(s) to display the full node tree in the scene tree. Then, in the scene tree, you can see a pose (relative to any other node), velocity, joint properties...

Very cool! Didn't know that.
However, my question was more generic: can you please point me to the code where webots computes transformations? It's a big codebase and I'm yet to fully grasp it.

**Comment by omichel** [comment]
> I am aware of that. But how can I do that programmatically? I tried to getField("Position").getSFVec3f() but it doesn't work.

You should use [node.getPosition()](https://www.cyberbotics.com/doc/reference/supervisor?tab-language=python#wb_supervisor_node_get_position) instead.

> Also, I am not understanding how to get the list of attributes for a node/object.

You should use [node.getNumberOfFields()](https://www.cyberbotics.com/doc/reference/supervisor?tab-language=python#wb_supervisor_node_get_number_of_fields) and [node.getFieldByIndex()](https://www.cyberbotics.com/doc/reference/supervisor?tab-language=python#wb_supervisor_node_get_field_by_index) to browse all the fields of a node. See also [field.getName()](https://www.cyberbotics.com/doc/reference/supervisor?tab-language=python#wb_supervisor_field_get_name) and [field.getType()](https://www.cyberbotics.com/doc/reference/supervisor?tab-language=python#wb_supervisor_field_get_type).

> However, my question was more generic: can you please point me to the code where webots computes transformations? It's a big codebase and I'm yet to fully grasp it.

Maybe you can start [here](https://github.com/cyberbotics/webots/blob/master/src/webots/nodes/WbTransform.hpp).

**Reply by alecive** [reply]
Wonderful, thanks!
